File: src/0946-Validate-Stack-Sequences/0946.js
var validateStackSequences = function(pushed, popped) {
    let i = 0, j = 0;

    for (let num of pushed) {
        pushed[i] = num;
        i++;
        while (i > 0 && pushed[i - 1] == popped[j]) i--, j++;
    }
    
    return i == 0;
};

File: src/0946-Validate-Stack-Sequences/0946.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        int i = 0, j = 0;

        for (int num : pushed) {
            pushed[i] = num;
            i++;
            while (i > 0 && pushed[i - 1] == popped[j]) {
                i--;
                j++;
            }
        }
        
        return i == 0;
    }
}